dev-clean/headers/llsc.hpp

```diff
@@ -553,14 +573,202 @@ class PC {
+      return name;
+    }
+    immer::flex_vector<PtrVal> read_at(size_t pos, size_t length) const {
+      return content.drop(pos).take(length);
+    }
+    immer::flex_vector<PtrVal> get_content() const {
+      return read_at(0, content.size());
+    }
+};
+
+// return a symbolic file with size bytes
+inline File make_SymFile(std::string name, size_t size) {
+  immer::flex_vector<PtrVal> content;
+  for (int i = 0; i < size; i++) {
+    content = content.push_back(make_SymV(std::string("FILE_") + name + std::string("_BYTE_") + std::to_string(i)));
+  }
+  return File(name, content);
+};
+
+/* TODO: what is the rule about the lowest file descriptor guarantee?
+ * How do we model that rule? Is there a usecase for that?
+ * Use a particular data structure? <2021-10-12, David Deng> */
+
+// An opened file
+struct Stream {
+  private:
+    File file;
+    int mode; // a combination of O_RDONLY, O_WRONLY, O_RDWR, etc.
+    size_t cursor;
+    status_t status {status_good_bit};
+  public:
+    // error handling
+    static const status_t status_good_bit = 0x0;
+    static const status_t status_seek_fail_bit = 0x1 << 1;
+    inline status_t get_status() const { return status; }
+    inline Stream set_status(status_t s) { return Stream(file, mode, cursor, status | s); }
+    inline Stream clear_status(status_t s = status_good_bit) { return Stream(file, mode, cursor, s); }
+    inline bool status_good() const { return ~(status ^ status_good_bit); } // status must match status_good_bit
+    inline bool status_seek_fail() const { return status & status_seek_fail_bit; }
+
+    Stream(File file):
+      file(file),
+      mode(O_RDONLY),
+      cursor(0) {}
+    Stream(File file, int mode):
+      file(file),
+      mode(mode),
+      cursor(0) {}
+    Stream(File file, int mode, size_t cursor):
+      file(file),
+      mode(mode),
+      cursor(cursor) {}
+    Stream(File file, int mode, size_t cursor, status_t status):
+      file(file),
+      mode(mode),
+      cursor(cursor),
+      status(status) {}
+    Stream seek_start(long offset) {
+      if (offset < 0) return set_status(status_seek_fail_bit);
+      return Stream(file, mode, offset);
+    }
+    Stream seek_end(long offset) {
+      long new_cursor = file.get_size() + offset;
+      if (new_cursor < 0) return set_status(status_seek_fail_bit);
+      return Stream(file, mode, new_cursor);
+    }
+    Stream seek_cur(long offset) {
+      long new_cursor = cursor + offset;
+      if (new_cursor < 0) return set_status(status_seek_fail_bit);
+      return Stream(file, mode, new_cursor);
+    }
+    size_t get_cursor() const { return cursor; }
+};
+
+class FS {
+  private:
+    /* TODO: How does immer handle memory leak if map.find returns a raw pointer? <2021-10-12, David Deng> */
+    /* NOTE: Use at instead of find for immer::map <2021-10-12, David Deng> */
+
+    immer::map<Fd, Stream> open_files;
+    immer::map<std::string, File> files;
+    Fd next_fd;
+    Fd open_fd; // set by open()
```

Comment:
It is basically used as a way for the model (`FS`) to report the file descriptor of the last opened file (E.g. when someone calls `open_file()`), if that makes any sense.

Reply:
I see. It seems better to rename it as `last_opened_fd` (and `open_files` -> `opened_files`).

I'm also a bit concerned about whether this invariant "open_fd is the last opened file" is properly maintained: when `close_file` is called, it does not change `open_fd`. If we call `close_file` right after `open_file` with the same filename, then this internal `close_file` status is not valid anymore.

Reply:
I agree.
- `last_opened_fd` makes sense.
- `opened_files` is also better.

As for `open_fd`, the design is that we should only check it after calling `fs.open_file()`, so that we can know how to refer to the file we just opened. It doesn't need to be synchronized to the "second last opened file" or anything like that if we call other methods such as `close_file` later on that file.

Reply:
I see -- would it make sense to return a pair that directly contains that lastly opened fd?

The other concern is whether it is necessary to use a functional/pure API of `File` and `Stream`, it seems any changes of `File`/`Stream` returns a new object, and you need to put it back into `open_files`/`files` of `FS` as well (otherwise calling `get_stream` on the same `fd` would lead to inconsistency), which could be cumbersome. 
I think, internally at `File`/`Stream`, it could be more straightforward to just mutate their states.

Reply:
It is definitely an option to return a pair, but I wasn't sure about the performance. Do you think it's better than setting the status?

As for the immutable style, I also found it quite tedious to overload constructors. But why do we use the immer library at all? Under what conditions do we want to use the functional API? Would it cause race conditions on parallel threads if we use mutable objects to model the file system?

Reply:
> It is definitely an option to return a pair, but I wasn't sure about the performance. Do you think it's better than setting the status?

I wouldn't think there is a big overhead of using `std::pair`, which is a polymorphic template of struct. Usually, it ends up with a struct with the same memory footprint. 

> As for the immutable style, I also found it quite tedious to overload constructors. But why do we use the immer library at all? Under what conditions do we want to use the functional API?

Not sure the immutable style is related to overloading constructors... The generated code observes an immutable/pure interface of state (ie `SS`), but it doesn't mean the internal implementation of the state has to be pure as well. For example, the modeling of the file system is completely encapsulated and not exposed to generated code, and the only way to use it is through our own customized `open`/`close`/syscalls functions. 

> Would it cause race conditions on parallel threads if we use mutable objects to model the file system?

We certainly want to avoid race conditions, but it ultimately depends on how you write the code. If two threads operate on different `FS` objects and they do not share aliases, then it should be safe.

Reply:
FYI, `PairV` is already implemented in the concolic executor [backend](https://github.com/Kraks/sai/blob/master/dev-clean/headers/conc_global_solver.hpp#L262) and [frontend](https://github.com/Kraks/sai/blob/master/dev-clean/src/main/scala/sai/conc/SymExeState.scala#L268) if you consider generate code to use pair.

Reply:
@Shangyint But `PairV` is supposed to represent pair values from the target program, and it should not be used to represent ordinary pairs in the backend implementation, right?

Reply:
Sorry I am a little lost. Is this pair value returned from an external function call?
Anyways you are right, `PairV` is only needed to represent values in the target program. Ordinary pairs in the backend can just be represented as `std::pair`

Reply:
> I wouldn't think there is a big overhead of using std::pair, which is a polymorphic template of struct. Usually, it ends up with a struct with the same memory footprint.

That's good to know. However, I feel like it is mostly a syntactic difference between returning pair and setting a flag and don't see why one is superior than the other.

> Not sure the immutable style is related to overloading constructors... 

Yeah, other than overloading constructor, another way to "modify" immutable objects I see is to provide methods that returns copies of object, such as a `FS FS::set_flag(flag_t flag)` method.

> If two threads operate on different FS objects and they do not share aliases, then it should be safe.

Then I imagine we would need to enforce some constraints on when and what objects should be shared. For example, currently we are sharing objects as much as possible, and two threads can share the same `FS` object because they are immutable. If we make `FS` mutable, then we would have to enforce an explicitly copying the `FS` object at some point. For example,
- the `SS::get_fs` function always returns a (deep) copy of the stored `FS`, or
- the `SS::set_fs` function makes a deep copy of the `FS` object passed in the argument, or
- the client of `SS::get_fs` and `SS::set_fs` makes deep copies upon getting a `FS` instance before modifying it,
- ...

Do you think that would make things easier rather than more complicated?

Reply:
> Sorry I am a little lost. Is this pair value returned from an external function call?

Because we currently use immutable objects, every method that performs some modification to the immutable object has to return a new object containing the modification. The pair is supposed to be used for returning an extra value (e.g. whether the operation was successful) in addition to the modified object. Hope that helps!

Reply:
> That's good to know. However, I feel like it is mostly a syntactic difference between returning pair and setting a flag and don't see why one is superior than the other.

As you mention, "the design is that we should only check `fs.get_open_fd()` after calling `fs.open_file()`", however, this is not enforced by any means -- calling `get_open_fd` at any other time is essentially "undefined". So I would think returning the "just opened fd" together with `open_file()` is a better design, it is simply impossible to "get open fd" by other means, therefore the invariant of your design is kept.

> Then I imagine we would need to enforce some constraints on when and what objects should be shared. For example, currently we are sharing objects as much as possible, and two threads can share the same `FS` object because they are immutable. If we make `FS` mutable, then we would have to enforce an explicitly copying the `FS` object at some point. For example,
> 
> * the `SS::get_fs` function always returns a (deep) copy of the stored `FS`, or
> * the `SS::set_fs` function makes a deep copy of the `FS` object passed in the argument, or
> * the client of `SS::get_fs` and `SS::set_fs` makes deep copies upon getting a `FS` instance before modifying it,
> * ...

I'm a bit confused here. The current `SS::get_fs` and `SS::set_fs` already implicitly invoked the copy constructor of `FS`, right? Once the client gets a `FS` from `SS::get_fs`, any change to the returned `FS` has nothing to do with the `FS` in that `SS`.

What I mean by making use of mutable objects internally is something like this:

```
struct SS {
    FS fs;
    ...
};
struct FS {
    immer::map<...> fds;
    int flag;
    void set_flat(...) { flag = ... }
    void some_effectful_method() {
        fds = ...
        // even fds is immer::map, we can still change its value because it is not a const, so of course fds could be just std::map
    }
};
// in the *internal* code, we could simply get a reference to the FS of the current SS, 
// and perform implace updates
FS& fs = SS.get_FS();
fs.set_flag(...);
fs.some_effectful_method();
```

`FS` and `SS` do not hold any pointer of reference, once they are copied (by the copy constructor), their fields are also copied, therefore no sharing happens.

Reply:
> As you mention, "the design is that we should only check fs.get_open_fd() after calling fs.open_file()", however, this is not enforced by any means -- calling get_open_fd at any other time is essentially "undefined". So I would think returning the "just opened fd" together with open_file() is a better design, it is simply impossible to "get open fd" by other means, therefore the invariant of your design is kept.

Avoiding undefined behavior makes sense to me, but the interface seems a little inconsistent if for some methods (e.g. `open_file`) we return a `pair<SS, int>` and for some others we return just an `SS`. It might also be possible that we have some methods needing to report several status, in that case, do we want the return signature to be something like `pair<SS, pair<int, int>>`, or `pair<SS, vector<int>>`? Alternatively, we can also set a few flag variables in the `FS` objects, and let the "client" check the flags they need to check without making the return type signature inconsistent.

> FS and SS do not hold any pointer of reference, once they are copied (by the copy constructor), their fields are also copied, therefore no sharing happens.

I think you are right, since we are passing by value, the objects would be copied by default. Maybe I should do some redesign and refactoring of the interface, which would also solve the return value problem.

Reply:
Do you want the refactoring to be done in the current branch? Or can we merge this branch for now and I check out another branch for refactoring?

Reply:
>  the interface seems a little inconsistent if for some methods (e.g. open_file) we return a pair<SS, int> and for some others we return just an SS ...

That's a good point and also the reason that I mentioned those internal classes could just mutate their states/fields. Then `FS::open_file` could simply return an integer of fd.

Since we are now in the middle of experimenting with different header backends in master, let's make the refactor in this branch. Once that experiment is done (sometime next week), we could merge this branch.